src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -295,54 +383,141 @@ TaskSketcherElements::TaskSketcherElements(ViewProviderSketch *sketchView)
         boost::bind(&SketcherGui::TaskSketcherElements::slotElementsChanged, this));
 
     this->groupLayout()->addWidget(proxy);
 
-    ui->comboBoxElementFilter->setCurrentIndex(0);
-    ui->comboBoxModeFilter->setCurrentIndex(0);
 
-    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/Elements");
+    slotElementsChanged();
 
-    ui->autoSwitchBox->setChecked(hGrp->GetBool("Auto-switch to edge", true));
-    ui->namingBox->setChecked(hGrp->GetBool("Extended Naming", false));
+    // make filter items checkable
+    ui->listMultiFilter->blockSignals(true);
+    for (int i = 0; i < ui->listMultiFilter->count(); i++) {
+        QListWidgetItem* item = ui->listMultiFilter->item(i);
 
-    ui->comboBoxElementFilter->setEnabled(!isautoSwitchBoxChecked);
-    ui->comboBoxModeFilter->setEnabled(true);
+        item->setFlags(item->flags() | Qt::ItemIsUserCheckable);
 
-    slotElementsChanged();
+        item->setCheckState(Qt::Checked);
+    }
+    ui->listMultiFilter->setVisible(false);
+    ui->listMultiFilter->blockSignals(false);
+
+    this->installEventFilter(this);
+    ui->filterBox->installEventFilter(this);
+    ui->listMultiFilter->installEventFilter(this);
 }
 
 TaskSketcherElements::~TaskSketcherElements()
 {
-    try {
-        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/Elements");
-        hGrp->SetBool("Auto-switch to edge", ui->autoSwitchBox->isChecked());
-        hGrp->SetBool("Extended Naming", ui->namingBox->isChecked());
+    connectionElementsChanged.disconnect();
+}
+
+/* filter functions --------------------------------------------------- */
+
+void TaskSketcherElements::on_filterBox_stateChanged(int val){ 
+    if(ui->filterBox->checkState() == Qt::Checked)
+        ui->listMultiFilter->show();
+    else
+        ui->listMultiFilter->hide();
+
+    slotElementsChanged();
+}
+
+bool TaskSketcherElements::eventFilter(QObject* obj, QEvent* event)
+{
+    if (obj == (QObject*)ui->filterBox && event->type() == QEvent::Enter && ui->filterBox->checkState() == Qt::Checked) {
+        ui->listMultiFilter->show();
     }
-    catch (const Base::Exception&) {
+    else if (obj == (QObject*)ui->listMultiFilter && event->type() == QEvent::Leave) {
+        ui->listMultiFilter->hide();
     }
+    else if (obj == this && event->type() == QEvent::Leave) {
+        ui->listMultiFilter->hide();
+    }
+    return QWidget::eventFilter(obj, event);
+}
 
-    connectionElementsChanged.disconnect();
+void TaskSketcherElements::on_listMultiFilter_itemChanged(QListWidgetItem* item)
+{
+    ui->listMultiFilter->blockSignals(true);
+
+    if (item == ui->listMultiFilter->item(3)) { //3 is 'All geos'
+        for (int i = 4; i < ui->listMultiFilter->count(); i++) {
+            ui->listMultiFilter->item(i)->setCheckState(item->checkState());
+        }
+    }

```

Comment:
Pretty horrible. Maybe do same thing as `filterAggregates` in constraints ?

Reply:
I disagree. This is simple and understandable. filterAggregates is hard to understand. And it is absolutely not needed as the filter here is very simple.
This would introduce unnecessary complexity.

Reply:
This is simple now (and even, one has to go to the UI file to see what is what). The problem is maintaining the code is one want to add another line to the filter. It may just break everything.

Reply:
If someone wants to add another line to the filter, he will have to understand the code structure in all cases. If it's a filterAggregates he will also have to understand the code structure, but it's much more complex to understand.